pgrx/src/fcall.rs

```diff
@@ -0,0 +1,448 @@
+    // we must return `None` now and not call the function.  Passing a NULL argument to a STRICT
+    // function will likely crash Postgres
+    let isstrict = pg_proc.proisstrict();
+    if null && isstrict {
+        return Ok(None);
+    }
+
+    //
+    // The following code is Postgres-version specific.  Right now, it's compatible with v12+
+    // v11 will need a different implementation.
+    //
+    // NB:  Which I don't want to do since it EOLs in November 2023
+    //
+
+    unsafe {
+        // construct a stack-allocated `FmgrInfo` instance
+        let mut flinfo = pg_sys::FmgrInfo::default();
+
+        // SAFETY:  we just allocated `flinfo`.  Whatever objects `fmgr_info` may allocate
+        // are allocated in `CurrentMemoryContext`, which is fine as this `flinfo` doesn't live longer
+        // than this stack frame anyway.
+        pg_sys::fmgr_info(func_oid, &mut flinfo);
+
+        // heap allocate a `FunctionCallInfoBaseData` properly sized so there's enough room for `nargs` arguments
+        //
+        // SAFETY: we allocate enough zeroed space for the base FunctionCallInfoBaseData *plus* the number of arguments
+        // we have, and we've asserted that we have the correct number of arguments
+        assert_eq!(nargs, pg_proc.pronargs());
+        let fcinfo = pg_sys::palloc0(
+            std::mem::size_of::<pg_sys::FunctionCallInfoBaseData>()
+                + std::mem::size_of::<pg_sys::NullableDatum>() * nargs,
+        ) as *mut pg_sys::FunctionCallInfoBaseData;
+
+        // initialize it
+        // SAFETY: we just palloc'd the `fcinfo` instance so it's de-referencable
+        let fcinfo_ref = &mut *fcinfo;
+        fcinfo_ref.flinfo = &mut flinfo;
+        fcinfo_ref.fncollation = collation;
+        fcinfo_ref.context = std::ptr::null_mut();
+        fcinfo_ref.resultinfo = std::ptr::null_mut();
+        fcinfo_ref.isnull = false;
+        fcinfo_ref.nargs = nargs as _;
+
+        // setup the argument array
+        // SAFETY:  `fcinfo_ref.args` is the over-allocated space we palloc0'd above.  it's an array
+        // of `nargs` `NulalbleDatum` instances.
+        let args_slice = fcinfo_ref.args.as_mut_slice(nargs);
+        for (i, datum) in arg_datums.into_iter().enumerate() {
+            assert!(!isstrict || (isstrict && datum.is_some())); // no NULL datums if this function is STRICT
+
+            let arg = &mut args_slice[i];
+            (arg.value, arg.isnull) =
+                datum.map(|d| (d, false)).unwrap_or_else(|| (pg_sys::Datum::from(0), true));
+        }
+
+        // call the function
+        // SAFETY: `flinfo` was create for us by `fmgr_info` above and `fn_addr` would have been properly set
+        let func = *(*fcinfo_ref.flinfo)
+            .fn_addr
+            .as_ref()
+            .expect("function initialization problem: fn_addr not set");
+        let result_datum = func(fcinfo);
```

Comment:
Just thought of this, but this is (potentially) an external call into a C postgres function, so if it errors it will longjmp, no? So I think we need a [`pg_guard_ffi_boundary`](https://github.com/pgcentralfoundation/pgrx/blob/develop/pgrx-pg-sys/src/submodules/ffi.rs#L92) around it, unless I'm mistaken

Reply:
yeah, great catch!  Didn't even occur to me.  :(

Reply:
Fixed up, with a test, in https://github.com/pgcentralfoundation/pgrx/pull/1279/commits/7616a6685c27f444afde64c6219b1ca2453c4a3d